synapse/lib/auth.py

```diff
@@ -1267,6 +1286,104 @@ async def tryPasswd(self, passwd, nexs=True):
+            failures.append(f'Password must be at least {minlen} characters.')
+
+        if minlen and passwd is None:
+            # Set password to empty string so we get the rest of the failure info
+            passwd = ''
+
+        if passwd is None:
+            return
+
+        allvalid = []
+
+        # Check uppercase
+        count = complexity.get('upper:count')
+        valid = complexity.get('upper:valid')
+        allvalid.append(valid)
+
+        if count and (found := len([k for k in passwd if k in valid])) < count:
+            failures.append(f'Password must contain at least {count} uppercase characters, {found} found.')
+
+        # Check lowercase
+        count = complexity.get('lower:count')
+        valid = complexity.get('lower:valid')
+        allvalid.append(valid)
+
+        if count and (found := len([k for k in passwd if k in valid])) < count:
+            failures.append(f'Password must contain at least {count} lowercase characters, {found} found.')
+
+        # Check special
+        count = complexity.get('special:count')
+        valid = complexity.get('special:valid')
+        allvalid.append(valid)
+
+        if count and (found := len([k for k in passwd if k in valid])) < count:
+            failures.append(f'Password must contain at least {count} special characters, {found} found.')
+
+        # Check numbers
+        count = complexity.get('number:count')
+        valid = complexity.get('number:valid')
+        allvalid.append(valid)
+
+        if count and (found := len([k for k in passwd if k in valid])) < count:
+            failures.append(f'Password must contain at least {count} digit characters, {found} found.')
+
+        allvalid = ''.join(allvalid)
+        if (invalid := set(passwd) - set(allvalid)):
+            failures.append(f'Password contains invalid characters: {sorted(list(invalid))}')
+
+        # Check sequences
+        seqlen = complexity.get('sequences')
+        if seqlen > 1:
+            # Convert each character to it's ordinal value so we can look for
+            # forward and reverse sequences in windows of seqlen. Doing it this
+            # way allows us to easily check unicode sequences too.
+            passb = [ord(k) for k in passwd]
+            for offs in range(len(passwd) - (seqlen - 1)):
+                curv = passb[offs]
+                fseq = list(range(curv, curv + seqlen))
+                rseq = list(range(curv, curv - seqlen, -1))
+                window = passb[offs:offs + seqlen]
+                if window == fseq or window == rseq:
+                    failures.append(f'Password must not contain forward/reverse sequences longer than {seqlen} characters.')
+                    break
+
+        # Check for previous password reuse
+        prevvalu = self.auth.policy.get('previous')
+        if prevvalu:
+            previous = self.info.get('policy:previous', ())
+            for prevshad in previous:
```

Comment:
Does this mean that we are required to shadow new passwords using the same salt?

Reply:
No, we just iterate over the old shadows using their salt to check the new password against the old passwords. A new shadow is created for each password.